src/persist/clients.ts

```diff
@@ -72,83 +78,151 @@ export async function getClient(
   return clients.get(id);
 }
 
-export async function setClients(
-  clients: ClientMap,
-  dagWrite: dag.Write,
-): Promise<void> {
-  const chunkData = clientMapToChunkData(clients, dagWrite);
-  const chunk = dagWrite.createChunk(
-    chunkData,
-    Array.from(clients.values(), client => client.headHash),
-  );
-  await Promise.all([
-    dagWrite.putChunk(chunk),
-    dagWrite.setHead(CLIENTS_HEAD, chunk.hash),
-  ]);
-}
-
-export async function setClient(
-  id: ClientID,
-  client: Client,
-  dagWrite: dag.Write,
-): Promise<void> {
-  const clients = await getClients(dagWrite);
-  clients.set(id, client);
-  return setClients(clients, dagWrite);
-}
-
 export async function initClient(
-  dagWrite: dag.Write,
+  dagStore: dag.Store,
 ): Promise<[ClientID, Client]> {
-  const clients = await getClients(dagWrite);
   const newClientID = makeUuid();
-  let bootstrapClient;
-  for (const client of clients.values()) {
-    if (
-      !bootstrapClient ||
-      bootstrapClient.heartbeatTimestampMs < client.heartbeatTimestampMs
-    ) {
-      bootstrapClient = client;
+  const updatedClients = await updateClients(async clients => {
+    let bootstrapClient: Client | undefined;
+    for (const client of clients.values()) {
+      if (
+        !bootstrapClient ||
+        bootstrapClient.heartbeatTimestampMs < client.heartbeatTimestampMs
+      ) {
+        bootstrapClient = client;
+      }
     }
-  }
 
-  let newClientCommit;
-  if (bootstrapClient) {
-    const bootstrapCommit = await db.baseSnapshot(
-      bootstrapClient.headHash,
+    const newClientCommitData = await dagStore.withRead(async dagRead => {
+      if (bootstrapClient) {
+        const bootstrapCommit = await db.baseSnapshot(
+          bootstrapClient.headHash,
+          dagRead,
+        );
+        // Copy the snapshot with one change: set last mutation id to 0.  Replicache
+        // server implementations expect new client ids to start with last mutation id 0.
+        // If a server sees a new client id with a non-0 last mutation id, it may conclude
+        // this is a very old client whose state has been garbage collected on the server.
+        return newSnapshotCommitData(
+          bootstrapCommit.meta.basisHash,
+          0 /* lastMutationID */,
+          bootstrapCommit.meta.cookieJSON,
+          bootstrapCommit.valueHash,
+          bootstrapCommit.indexes,
+        );
+      } else {
+        // No existing snapshot to bootstrap from. Create empty snapshot.
+        // TODO This empty btree write is a problem...
+        //const emptyBTreeHash = await new BTreeWrite(dagWrite).flush();
```

Comment:
This is the one significant difficulty remaining.  I need some way of getting the has of the empty btree, and the chunks for the empty btree to return as chunksToPut.

Reply:
fixed with Erik's approach.